### PR TITLE
Native related code refactor

### PR DESCRIPTION
**Pull request by pm-harshad-mane**
## Native related code refactor and Supporting Multi-format requests
- [X] Code style update (formatting, local variables)
- [X] Refactoring (no functional changes, no api changes)
- [X] New feature

## Description of change
- Native related code refactor
- Supporting Multi-format request

### CODE REVIEW

**Comment by pm-harshad-mane** [comment]
I think there is some issue with CI, I just removed a few comments and now CI is failing.
Checked on my machine, lint and test is working fine.

**Comment by pm-harshad-mane** [comment]
CI is working fine after updating the code from master

**Comment by pm-harshad-mane** [comment]
Hello @jsnellbaker ,

Please wait, we are trying to add some more changes in this branch. :-)

File: modules/pubmaticBidAdapter.js [hunk 1]
```diff
@@ -220,26 +172,37 @@ function _parseAdSlot(bid) {
...
+    // delete bid.sizes;
+  } else {
+    if (bid.hasOwnProperty('mediaTypes') &&
+         bid.mediaTypes.hasOwnProperty(BANNER) &&
+          bid.mediaTypes.banner.hasOwnProperty('sizes')) {
+      bid.params.width = parseInt(bid.mediaTypes.banner.sizes[0][0]);
+      bid.params.height = parseInt(bid.mediaTypes.banner.sizes[0][1]);
+      bid.mediaTypes.banner.sizes = bid.mediaTypes.banner.sizes.splice(1, bid.mediaTypes.banner.sizes.length - 1);
+    } else {
+      utils.logWarn(BIDDER_CODE + ' error: Missing mediaTypes.banner in adSlot: ' + bid.params.adSlot);
```
Comment (pm-harshad-mane) [inline modules/pubmaticBidAdapter.js]: use LOG_WARN_PREFIX

File: modules/pubmaticBidAdapter.js [hunk 2]
```diff
@@ -220,26 +172,37 @@ function _parseAdSlot(bid) {
...
     if (splits.length != 2) {
       utils.logWarn(LOG_WARN_PREFIX + 'AdSlot Error: adSlot not in required format');
       return;
     }
     bid.params.width = parseInt(splits[0]);
     bid.params.height = parseInt(splits[1]);
-    delete bid.sizes;
-  } else if (sizesArrayExists) {
+    // delete bid.sizes;
+  } else {
```
Comment (pm-harshad-mane) [inline modules/pubmaticBidAdapter.js]: can we use if - elseif - else ?

File: modules/pubmaticBidAdapter.js [hunk 3]
```diff
@@ -220,26 +172,37 @@ function _parseAdSlot(bid) {
...
     splits = splits[1].split('x');
     if (splits.length != 2) {
       utils.logWarn(LOG_WARN_PREFIX + 'AdSlot Error: adSlot not in required format');
       return;
     }
     bid.params.width = parseInt(splits[0]);
     bid.params.height = parseInt(splits[1]);
-    delete bid.sizes;
-  } else if (sizesArrayExists) {
+    // delete bid.sizes;
```
Comment (pm-harshad-mane) [inline modules/pubmaticBidAdapter.js]: can we remove commented code?

File: modules/pubmaticBidAdapter.js [hunk 4]
```diff
@@ -587,11 +436,76 @@ function _createNativeRequest(params) {
...
+      format = [];
+      sizes.forEach(function (size) {
+        format.push({ w: size[0], h: size[1] });
+      });
+      bannerObj.format = format;
+    }
+    bannerObj.pos = 0;
+    bannerObj.topframe = utils.inIframe() ? 0 : 1;
+  } else {
+    utils.logWarn(BIDDER_CODE + 'Error: mediaTypes.banner.size missing for adunit: ' + bid.params.adUnit + '. Ignoring the banner impression in the adunit.');
```
Comment (pm-harshad-mane) [inline modules/pubmaticBidAdapter.js]: use LOG_WARN_PREFIX

File: modules/pubmaticBidAdapter.js [hunk 5]
```diff
@@ -604,42 +518,42 @@ function _createImpressionObject(bid, conf) {
...
-    }
-    if (bid.params.video.hasOwnProperty('skippable')) {
-      videoObj.ext = {
-        'video_skippable': bid.params.video.skippable ? 1 : 0
+  if (bid.hasOwnProperty('mediaTypes')) {
+    for (mediaTypes in bid.mediaTypes) {
+      switch (mediaTypes) {
+        case BANNER:
+          bannerObj = _createBannerRequest(bid);
+          if (bannerObj !== undefined) {
```
Comment (pm-harshad-mane) [inline modules/pubmaticBidAdapter.js]: use const UNDEFINED

File: modules/pubmaticBidAdapter.js [hunk 6]
```diff
@@ -710,6 +623,24 @@ function _handleEids(payload) {
...
+  } else if (adm.indexOf('span class="PubAPIAd"') >= 0) {
+    newBid.mediaType = BANNER;
+  }
+  try {
+    admStr = JSON.parse(adm.replace(/\\/g, ''));
+    if (admStr && admStr.native) {
+      newBid.mediaType = NATIVE;
+    }
+  } catch (e) {
+
```
Comment (pm-harshad-mane) [inline modules/pubmaticBidAdapter.js]: should we put some warning here?

File: modules/pubmaticBidAdapter.js (hunk 1 above)
Comment (pm-harshad-mane) [reply modules/pubmaticBidAdapter.js]: fixed

File: modules/pubmaticBidAdapter.js (hunk 5 above)
Comment (pm-harshad-mane) [reply modules/pubmaticBidAdapter.js]: done

File: modules/pubmaticBidAdapter.js (hunk 6 above)
Comment (pm-harshad-mane) [reply modules/pubmaticBidAdapter.js]: done

File: modules/pubmaticBidAdapter.js (hunk 4 above)
Comment (pm-harshad-mane) [reply modules/pubmaticBidAdapter.js]: done

File: modules/pubmaticBidAdapter.js [hunk 7]
```diff
@@ -306,35 +264,41 @@ function _createOrtbTemplate(conf) {
   };
 }
 
-// similar functionality as parseSlotParam. Check if the 2 functions can be clubbed.
 function _checkParamDataType(key, value, datatype) {
   var errMsg = 'PubMatic: Ignoring param key: ' + key + ', expects ' + datatype + ', found ' + typeof value;
```
Comment (pm-harshad-mane) [inline modules/pubmaticBidAdapter.js]: need to remove PubMatic from message

File: modules/pubmaticBidAdapter.js [hunk 8]
```diff
@@ -710,6 +618,24 @@ function _handleEids(payload) {
     payload.user.eids = eids;
   }
 }
+// TODO: need a better way to identify mediaType from response.
+function _checkMediaType(adm, newBid) {
+  // Create a regex here to check the strings
```
Comment (pm-harshad-mane) [inline modules/pubmaticBidAdapter.js]: todo: use regex for vast

File: modules/pubmaticBidAdapter.js (hunk 3 above)
Comment (pm-manasi-moghe) [reply modules/pubmaticBidAdapter.js]: done

File: modules/pubmaticBidAdapter.js (hunk 2 above)
Comment (pm-manasi-moghe) [reply modules/pubmaticBidAdapter.js]: done

File: modules/pubmaticBidAdapter.js (hunk 7 above)
Comment (pm-manasi-moghe) [reply modules/pubmaticBidAdapter.js]: done

File: modules/pubmaticBidAdapter.js (hunk 8 above)
Comment (pm-manasi-moghe) [reply modules/pubmaticBidAdapter.js]: done

Comment (pm-harshad-mane) [reply modules/pubmaticBidAdapter.js]: Hello @pm-manasi-moghe ,
Let's keep this comment unresolved until we merge the changes into this branch.

File: modules/pubmaticBidAdapter.js (hunk 3 above)
Comment (pm-harshad-mane) [reply modules/pubmaticBidAdapter.js]: Hello @pm-manasi-moghe ,
Let's keep this comment unresolved until we merge the changes into this branch.

**Comment by jsnellbaker** [comment]
@pm-harshad-mane @pm-manasi-moghe Just to clarify - are there still changes pending?  I saw the comments regarding the commented code, so I wanted to confirm the overall status.

**Comment by pm-harshad-mane** [comment]
Hello @jsnellbaker ,
Sorry for late response.
We are now ready with the changes.
Can you please review?

**Comment by pm-harshad-mane** [comment]
Thank you @jsnellbaker 👍

**Pull request merged**